### ISSUE TITLE
Numpy dev: 2 failing tests (string aggregation suddenly works)

### ISSUE THREAD

**Issue by neutrinoceros** [astropy/astropy#16211]
[example logs](https://github.com/astropy/astropy/actions/runs/8322042150/job/22769210629)

Reprod:
```shell
pytest astropy/table/tests/test_groups.py::test_table_aggregate
```
log extract
```
E       Failed: DID NOT WARN. No warnings of type (<class 'astropy.utils.exceptions.AstropyUserWarning'>,) were emitted.
E        Emitted warnings: [ UserWarning('Warning: converting a masked element to nan.'),
E         UserWarning('Warning: converting a masked element to nan.')].
```
What seems to be happening is that we get both expected warning *but* the second one doesn't have the expected category and is a plain `UserWarning` instead of `AstropyUserWarning`

This bisects to https://github.com/numpy/numpy/pull/26012, which was backported to numpy 2.0.x in  [numpy/numpy@`691514a` (#26047)](https://github.com/numpy/numpy/pull/26047/commits/691514a3dd35657279a77650d051c547ab10f13f) .

I'll dig some more to see if we should (and can) fix this on our side, and will report otherwise.

**Comment by neutrinoceros** [comment]
The following patch allows the test to go on (I'm not saying that it's correct)
```patch
diff --git a/astropy/table/tests/test_groups.py b/astropy/table/tests/test_groups.py
index e29c6d9b8a..b1723939e1 100644
--- a/astropy/table/tests/test_groups.py
+++ b/astropy/table/tests/test_groups.py
@@ -11,7 +11,7 @@ from astropy.table import Column, NdarrayMixin, QTable, Table, table_helpers, un
 from astropy.tests.helper import PYTEST_LT_8_0
 from astropy.time import Time
 from astropy.utils.exceptions import AstropyUserWarning
-
+from astropy.utils.compat import NUMPY_LT_2_0
 
 def sort_eq(list1, list2):
     return sorted(list1) == sorted(list2)
@@ -471,7 +471,7 @@ def test_table_aggregate(T1):
     t1m["d"].mask[4:6] = True
     tg = t1m.group_by("a")
 
-    if PYTEST_LT_8_0:
+    if PYTEST_LT_8_0 or not NUMPY_LT_2_0:
         ctx = nullcontext()
     else:
         ctx = pytest.warns(AstropyUserWarning, match="Cannot aggregate column")
```

... into the next problem 
https://github.com/astropy/astropy/blob/838979944cd62ef3be4e34991341f58742b2744a/astropy/table/tests/test_groups.py#L482-L489

what we actually get is
```
[' a   b    c    d    q  ',
 '                    m  ',
 '--- ---- ---- ---- ----',
 '  0    a  nan  nan  4.0',
 '  1  \x00\x00a  3.0 13.0 18.0',
 '  2 \x00\x00\x00a 22.0  6.0  6.0']
 ```
 
Seems to me that the particular behavior that is tested here relies on `np.sum` *not* being supported for string arrays, which isn't true in numpy 2.0 !
Basically, this boils down to

```python
>>> import numpy as np
>>> arr = np.array(["a", "b", "c", "d", "e"])
>>> np.sum(arr)
```

which tracebacks in numpy 1.x
```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/tmp/.venv/lib/python3.12/site-packages/numpy/core/fromnumeric.py", line 2313, in sum
    return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/.venv/lib/python3.12/site-packages/numpy/core/fromnumeric.py", line 88, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
numpy.core._exceptions._UFuncNoLoopError: ufunc 'add' did not contain a loop with signature matching types (dtype('<U1'), dtype('<U1')) -> None
```
while numpy 2.0 is happy to sum strings:
```python
'\x00\x00\x00\x00e'
```
though I'm not sure that this particular result is expected (a lot of data seems to actually be lost). I'll try to study the breaking change more closely before I decide whether to report upstream or not.

**Comment by neutrinoceros** [comment]
I think there's an actual bug, reported upstream as https://github.com/numpy/numpy/issues/26062
Note that our test might still require some adjustments when the upstream discussion is converged.

**Comment by neutrinoceros** [comment]
Reading the test more carefully, I'm not excluding that there's an actual bug in (or adjacent to) `TableGroup` that the change in numpy is merely revealing.

**Comment by neutrinoceros** [comment]
Hum, `str` typed columns being ignored in aggregations is actually not only tested but also documented https://docs.astropy.org/en/latest/table/operations.html#aggregation. I think we should just follow along numpy as we're already doing in other places where doing so is technically an API break (see #16167). If nothing else, I think this warrants a mention in #16200.

**Comment by mhvk** [comment]
Thanks for the sleuthing! The documentation suggests that indeed we did not support string aggregation because string summing was not supported (and `mean()` will continue to not be supported!). From the numpy discussion, it sounds like regular string dtype will still not support addition, but the new `StringDType` will. I'm certainly fine with letting that work in tables too.

**Comment by neutrinoceros** [comment]
> From the numpy discussion, it sounds like regular string dtype will still not support addition, but the new StringDType will. 

This would mean that we don't need to fix anything in astropy for the time being, right ?

> I'm certainly fine with letting that work in tables too.

Good ! As you said, `np.mean` won't be supported any way so, with that in mind, I think the existing phrasing in our docs is actually future proof, so again, nothing to do there !

**Comment by pllim** [comment]
tl;dr -- We have to handle *some* of the failures or they will all magically go away soon?

**Comment by neutrinoceros** [comment]
My understanding is that numpy dev silently promoting plain Python strings to a new dtype is considered a bug so I expect it will be fixed upstream and that we won't need to do anything more on our side. I'll keep this issue updated whichever direction is actually taken !

**Comment by mhvk** [comment]
New fix at https://github.com/numpy/numpy/pull/26090

**Comment by neutrinoceros** [comment]
Almost there. I once again ran our test suite against numpy dev (main) and it's now all green locally. However, I'd like to keep this issue open until the fix is actually backported to 2.0.x !

**Comment by pllim** [comment]
> backported to 2.0.x

But how do we test it? Would they release another beta?

**Comment by mhvk** [comment]
I think we should just close this issue once the fix has hit the nightlies. Then for `rc1` we can see whether there still is a problem (by that point, we probably should adjust our own workflows to run 2.0.0rc*`  as well).

**Comment by neutrinoceros** [comment]
Sounds like a plan. As long as we agree that we need to be testing the 2.0.x branch too (preferably just pre-releases), I'm fine with closing this sooner !

**Comment by neutrinoceros** [comment]
Backport PR is at https://github.com/numpy/numpy/pull/26111

**Comment by neutrinoceros** [comment]
Ok so devdeps with nightlies is indeed fixed now. *However* astropy main is still not compatible with `numpy@maintenance/v2.0.x`. I think the required fix is on our side this time, so I'll issue a PR.

**Comment by neutrinoceros** [comment]
For posterity, the incompatibility I was mentioning is fixed in #16229, is the only one I found, and isn't directly related to the present issue, so closing this now !
